components/director/pkg/operation/api.go

```diff
@@ -0,0 +1,152 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package operation
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	"github.com/kyma-incubator/compass/components/director/internal/model"
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/persistence"
+
+	"github.com/go-ozzo/ozzo-validation/is"
+
+	validation "github.com/go-ozzo/ozzo-validation"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/resource"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+)
+
+const ResourceIDParam = "resource_id"
+const ResourceTypeParam = "resource_type"
+
+// ResourceFetcherFunc defines a function which fetches a particular application by tenant and app ID
+type ResourceFetcherFunc func(ctx context.Context, tenant, id string) (*model.Application, error)
+
+// TenantLoaderFunc defines a function which fetches the tenant for a particular request
+type TenantLoaderFunc func(ctx context.Context) (string, error)
+
+type handler struct {
+	transact            persistence.Transactioner
+	resourceFetcherFunc ResourceFetcherFunc
+	tenantLoaderFunc    TenantLoaderFunc
+}
+
+// NewHandler creates a new handler struct associated with the Operations API
+func NewHandler(transact persistence.Transactioner, resourceFetcherFunc ResourceFetcherFunc, tenantLoaderFunc TenantLoaderFunc) *handler {
+	return &handler{
+		transact:            transact,
+		resourceFetcherFunc: resourceFetcherFunc,
+		tenantLoaderFunc:    tenantLoaderFunc,
+	}
+}
+
+// ServeHTTP handles the Operations API requests
+func (h *handler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	ctx := request.Context()
+
+	tenantID, err := h.tenantLoaderFunc(ctx)
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while retrieving tenant from context: %s", err.Error())
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to determine tenant for request"), http.StatusInternalServerError)
+		return
+	}
+
+	queryParams := request.URL.Query()
+
+	inputParams := struct {
+		ResourceID   string
+		ResourceType string
+	}{
+		ResourceID:   queryParams.Get(ResourceIDParam),
+		ResourceType: queryParams.Get(ResourceTypeParam),
+	}
+
+	log.C(ctx).Infof("Executing Operation API with resourceType: %s and resourceID: %s", inputParams.ResourceType, inputParams.ResourceID)
+
+	if err := validation.ValidateStruct(&inputParams,
```

Comment:
The op struct is internally created processed, I think we should expect it to be properly populated. This here is for the purpose to tell the client that his provided query params are invalid and he should correct them.

Reply:
Operation is still passed to scheduler and scheduler will has public Schedule method. Ideally a public method that has a contract with the outside world via such a struct should do a .Validate on it. So the best option would be to have such validate and call it in scheduler so that it can make sure it is given a valid op before sending to api server